src/Interpreters/ActionsVisitor.h

```diff
@@ -118,6 +118,7 @@ class ActionsMatcher
         bool no_makeset;
         bool only_consts;
         bool no_storage_or_local;
+        bool has_untuple = false;
```

Comment:
Another red flag, probably. I get the general idea, but the implementation looks like it is struggling to fit where it shouldn't. Maybe it is possible to move it higher than ActionsVisitor, somewhere around the place where the `*` is rewritten, and it will be more natural?

Reply:
The main reason it is in ActionsVisitor is that we need information about tuple type. Without it we even don't know how many columns we get from `untuple`.

I will try to remove this `CachedColumnName` cache